### PR TITLE
UI: Spectator UI refactor

### PR DESCRIPTION
**Pull request by mzegar**
Based off of branch https://github.com/TTTReborn/tttreborn/pull/140

### CODE REVIEW

File: code/global/Utils.cs [hunk 1]
```diff
@@ -70,6 +70,21 @@ public static List<TTTPlayer> GetConfirmedPlayers()
             return players;
         }
 
+        public static IEnumerable<Client> GetClientsSpectatingPlayerAndPlayer(TTTPlayer player)
+        {
+            List<Client> clients = new List<Client>{player.GetClientOwner()};
+
+            foreach (Client client in Client.All)
+            {
+                if (client.Pawn is TTTPlayer p && p.IsSpectatingPlayer && p.CurrentPlayer == player)
```
Comment (Alf21) [inline code/global/Utils.cs:79]:
```suggestion
            List<Client> clients = new();

            foreach (Client client in Client.All)
            {
                if (client.Pawn is TTTPlayer p && p.CurrentPlayer == player)
```

File: code/global/Utils.cs [hunk 2]
```diff
@@ -70,6 +70,21 @@ public static List<TTTPlayer> GetConfirmedPlayers()
             return players;
         }
 
+        public static IEnumerable<Client> GetClientsSpectatingPlayerAndPlayer(TTTPlayer player)
```
Comment (Alf21) [inline code/global/Utils.cs:73]: I'd remove `AndPlayer` as the player itself more or less observes the own client

File: code/player/TTTPlayer.cs [hunk 3]
```diff
@@ -20,6 +20,21 @@ public partial class TTTPlayer : Sandbox.Player
...
+        public TTTPlayer CurrentPlayer
+        {
+            get => _spectatingPlayer ?? this;
+            set
+            {
+                _spectatingPlayer = value;
+                Event.Run("tttreborn.player.spectating.change", this);
+            }
+        }
+        public bool IsSpectatingPlayer
```
Comment (Alf21) [inline code/player/TTTPlayer.cs:33]:
```suggestion

        public bool IsSpectatingPlayer
```

File: code/player/TTTPlayer.cs [hunk 4]
```diff
@@ -20,6 +20,21 @@ public partial class TTTPlayer : Sandbox.Player
...
+        {
+            get => CurrentPlayer != null && CurrentPlayer != this;
```
Comment (Alf21) [inline code/player/TTTPlayer.cs:35]:
```suggestion
            get => _spectatingPlayer != null;
```

File: code/player/TTTPlayer.cs [hunk 5]
```diff
@@ -20,6 +20,21 @@ public partial class TTTPlayer : Sandbox.Player
...
         public int Credits { get; set; } = 0;
 
+        private TTTPlayer _spectatingPlayer;
+        public TTTPlayer CurrentPlayer
+        {
+            get => _spectatingPlayer ?? this;
+            set
+            {
+                _spectatingPlayer = value;
+                Event.Run("tttreborn.player.spectating.change", this);
```
Comment (Alf21) [inline code/player/TTTPlayer.cs:30]:
```suggestion
                _spectatingPlayer = value == this ? null : value;

                Event.Run("tttreborn.player.spectating.change", this);
```

File: code/player/TTTPlayer.cs [hunk 6]
```diff
@@ -158,6 +171,7 @@ public override void Simulate(Client client)
             TickPlayerUse();
             TickPlayerDropCarriable();
             TickPlayerFlashlight();
+            TickAttemptInspectPlayerCorpse();
```
Comment (Alf21) [inline code/player/TTTPlayer.cs]: This wouldn't let FreeCam players inspect confirmed corpses

File: code/player/camera/ThirdPersonSpectateCamera.cs [hunk 7]
```diff
@@ -60,17 +58,17 @@ public override void Update()
...
+            if (Local.Pawn is not TTTPlayer player)
             {
                 return DefaultPosition;
             }
 
-            if (TargetPlayer == null || !TargetPlayer.IsValid())
+            if (!player.IsSpectatingPlayer)
             {
                 return DefaultPosition;
             }
```
Comment (Alf21) [inline code/player/camera/ThirdPersonSpectateCamera.cs:69]: Can be combined

File: code/ui/generalhud/nameplate/Nameplate.cs [hunk 8]
```diff
@@ -66,6 +65,7 @@ public Nameplate() : base()
             _nameLabel = _nameHolder.Add.Label("", "name");
 
             _damageIndicatorLabel = _labelHolder.Add.Label("", "damageIndicator");
+
```
Comment (Alf21) [inline code/ui/generalhud/nameplate/Nameplate.cs:68]: ?

File: code/ui/generalhud/nameplate/Nameplate.cs [hunk 9]
```diff
@@ -85,21 +85,19 @@ public override void Tick()
...
-            if (ObservedPlayer == null)
+            if (Local.Pawn is not TTTPlayer player)
             {
                 return;
             }
 
-            TTTPlayer player = Local.Pawn as TTTPlayer;
-
-            if (IsObserving && player.Camera is ThirdPersonSpectateCamera)
+            if (player.Camera is ThirdPersonCamera)
```
Comment (Alf21) [inline code/ui/generalhud/nameplate/Nameplate.cs:93]: Can be combined

File: code/ui/generalhud/nameplate/Nameplate.cs [hunk 10]
```diff
@@ -85,21 +85,19 @@ public override void Tick()
...
             {
                 return;
             }
 
             TraceResult trace = Trace.Ray(player.EyePos, player.EyePos + player.EyeRot.Forward * MAX_DRAW_DISTANCE)
+                .Ignore(player.ActiveChild)
                 .Ignore(player)
```
Comment (Alf21) [inline code/ui/generalhud/nameplate/Nameplate.cs:100]: Why changed this?

File: code/ui/generalhud/playerinfo/PlayerInfo.cs [hunk 11]
```diff
@@ -2,30 +2,29 @@
...
     {
-        public PlayerInfo() : base()
+        public PlayerInfo()
         {
             StyleSheet.Load("/ui/generalhud/playerinfo/PlayerInfo.scss");
 
             new RolePanel(this);
             new IndicatorsPanel(this);
-
-            AutoHideWithoutObservablePlayer = true;
```
Comment (Alf21) [inline code/ui/generalhud/playerinfo/PlayerInfo.cs:19]: Did you reimplemented this?

**Review by Alf21 — CHANGES_REQUESTED** [review]
All in all good changes, that will keep the adde logic easier

File: code/ui/generalhud/nameplate/Nameplate.cs [hunk 12]
```diff
@@ -85,18 +84,12 @@ public override void Tick()
...
-            }
-
-            if (player.Camera is ThirdPersonCamera)
+            if (Local.Pawn is not TTTPlayer player || player.Camera is ThirdPersonCamera)
             {
                 return;
             }
 
             TraceResult trace = Trace.Ray(player.EyePos, player.EyePos + player.EyeRot.Forward * MAX_DRAW_DISTANCE)
-                .Ignore(player.ActiveChild)
```
Comment (Alf21) [inline code/ui/generalhud/nameplate/Nameplate.cs]: Instead, we should ignore the observed player if the current player is observing

File: code/ui/generalhud/playerinfo/PlayerInfo.cs (hunk 11 above)
Comment (mzegar) [reply code/ui/generalhud/playerinfo/PlayerInfo.cs:19]: For the inventory selection, we don't need it, but for the player info we do. I'm open to suggestions on how you'd like to implement it.

For now I just did...
```
SetClass("hide", Local.Pawn is not TTTPlayer player || (player.IsSpectator && !player.IsSpectatingPlayer));
```

```
public bool IsSpectator
{
    // TODO: Change to be spectator class once FirstPersonSpectateCamera is added as well.
    get => (Camera is not FirstPersonCamera);
}
```

and then in that other PR I remember you adding a derived class for Spectating cameras, we can make use of that.

Comment (Alf21) [reply code/ui/generalhud/playerinfo/PlayerInfo.cs:19]: okay great, that's enough for me :D

File: code/ui/generalhud/nameplate/Nameplate.cs [hunk 13]
```diff
@@ -90,7 +90,7 @@ public override void Tick()
             }
 
             TraceResult trace = Trace.Ray(player.EyePos, player.EyePos + player.EyeRot.Forward * MAX_DRAW_DISTANCE)
-                .Ignore(player)
+                .Ignore(player.CurrentPlayer)
```
Comment (Alf21) [inline code/ui/generalhud/nameplate/Nameplate.cs:93]: I think we need 2 Ignores: Ignore the player and the player.CurrentPlayer if player.IsSpectating

File: code/ui/generalhud/playerinfo/PlayerInfo.cs [hunk 14]
```diff
@@ -8,6 +8,8 @@
 
 namespace TTTReborn.UI
 {
+    using Player = Sandbox.Player;
```
Comment (Alf21) [inline code/ui/generalhud/playerinfo/PlayerInfo.cs:11]: :D

**Review by Alf21 — APPROVED** [review]

**Pull request merged**